umi_tools/umi_methods.py

```diff
@@ -182,11 +182,38 @@ def get_read_position(read, soft_clip_threshold):
     return start, pos, is_spliced
 
 
-def get_bundles(insam, read_events,
-                ignore_umi=False, subset=None, quality_threshold=0,
-                paired=False, chrom=None, spliced=False, soft_clip_threshold=0,
-                per_contig=False, whole_contig=False, read_length=False,
-                detection_method="MAPQ", umi_getter=None, all_reads=False):
+def getMetaContig2contig(gene_transcript_map):
+    ''' '''
+    metacontig2contig = collections.defaultdict(set)
+    for line in U.openFile(gene_transcript_map, "r"):
+
+        if line.startswith("#"):
+            continue
+
+        if len(line.strip()) == 0:
+            break
+
+        gene, transcript = line.strip().split("\t")
+        metacontig2contig[gene].add(transcript)
+
+    return metacontig2contig
+
+
+def metafetcher(bamfile, metacontig2contig):
+    ''' return reads in order of metacontigs'''
+    for metacontig in metacontig2contig:
+        for contig in metacontig2contig[metacontig]:
+            for read in bamfile.fetch(contig):
+                read.tags += [('MC', metacontig)]
+                yield read
```

Comment:
This seems safer than updating the contig as the metacontig will be an unexpected value according to the sam header so could get problems with the reference id and reference nams etc.